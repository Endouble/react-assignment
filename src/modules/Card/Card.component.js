import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Card = (props) => {
    const noLink = (ev) => {
        ev.preventDefault();
        if (props.clickHandler) {
            props.clickHandler(props);
        }
    };
    const { name, imageUrl, set, number, artist, bioInfo, className } = props;
    const footStyles = `pokeCard__foot ${className && className}`;
    return (
        <StyledCard>
            <a className="pokeCard__item" href="/" onClick={noLink}>
                <h3 className="pokeCard__name">{number && `${number} - `} {name}</h3>
                <img className="pokeCard__img" alt={name} src={imageUrl} />
                <footer className={footStyles}>
                    <p>Set: {set}</p>
                    <p>Artist: {artist}</p>
                    {bioInfo && bioInfo}
                </footer>
            </a>
        </StyledCard>
    );
};

/*
    Card Styles
*/
const StyledCard = styled.article`
    padding: 10px;
    text-align: center;
    cursor: pointer;

    .pokeCard__item {
        text-decoration: none;
        color: inherit;
    }

    .pokeCard__name {
        margin: 5px;
        font-size: 20px;
        text-align: left;
        font-weight: bold;
    }

    .pokeCard__foot {
        background: #000850;
        color: white;
        font-size: 12px;
        margin: 0;
        padding: 10px;
        text-align: left;
        border-bottom-left-radius: 10px;
        border-bottom-right-radius: 10px;

        p {
            margin: 0 0 3px 0;
        }
    }

    .pokeCard__img {
        width: 100%;
        display: flex;
        background: #000850;
        border-top-left-radius: 10px;
        border-top-right-radius: 10px;
        min-height: 200px;
        margin-top: 0;
        transition: margin 0.25s ease;
    }

    &:hover, &:focus, &:active {
        .pokeCard__img {
            margin-top: -5px;
        }
    }
`;

/*
    Card propTypes
*/
Card.propTypes = {
    name: PropTypes.string,
    number: PropTypes.string,
    artist: PropTypes.string,
    imageUrl: PropTypes.string,
    set: PropTypes.string,
    clickHandler: PropTypes.func,
    bioInfo: PropTypes.node,
    className: PropTypes.string,
};
Card.defaultProps = {
    name: '',
    number: '',
    artist: '',
    imageUrl: '',
    set: '',
    clickHandler: () => null,
    bioInfo: null,
    className: '',
};

export default Card;
