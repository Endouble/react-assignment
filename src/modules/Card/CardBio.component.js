import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Card from './Card.component';

const CardBio = (props) => {
    const { hp, attacks, weaknesses } = props;
    const bioInfo = (
        <div>
            <p>HP: {hp}</p>
            <p>Attacks: {attacks.map(a => ` - ${a.name}`)} </p>
            <p>Weaknesses: {weaknesses.map(w => w.type)} </p>
        </div>
    );
    return (
        <StyledCardBio>
            <Card {...props} bioInfo={bioInfo} className="pokeBio" clickHandler={null} />
        </StyledCardBio>
    );
};

/*
    CardBio Styles
*/
const StyledCardBio = styled.article`
    .pokeBio {
        font-size: 16px;
        line-height: 1.5;
        background: none;
        text-align: center;
    }
`;

/*
    CardBio propTypes
*/
CardBio.propTypes = {
    hp: PropTypes.string,
    attacks: PropTypes.arrayOf(PropTypes.shape({})),
    weaknesses: PropTypes.arrayOf(PropTypes.shape({})),
};
CardBio.defaultProps = {
    hp: '',
    attacks: [],
    weaknesses: [],
};

export default CardBio;
