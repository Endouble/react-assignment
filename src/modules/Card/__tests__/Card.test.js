import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Card from '../';

configure({ adapter: new Adapter() });

const props = {
    name: 'testing card',
    imageUrl: 'testingUrl',
};

describe('Card', () => {
    it('renders correctly', () => {
        const component = shallow(<Card {...props} />);
        expect(component).toMatchSnapshot();
    });

    it('Has image', () => {
        const component = shallow(<Card {...props} />);
        const img = component.find('.pokeCard__img');
        expect(img.exists()).toEqual(true);
    });

    it('Has right alt image text', () => {
        const component = shallow(<Card {...props} />);
        const img = component.find('.pokeCard__img');
        expect(img.prop('alt')).toEqual(props.name);
    });
});
