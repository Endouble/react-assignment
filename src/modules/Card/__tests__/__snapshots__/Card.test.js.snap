// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders correctly 1`] = `
<styled.article>
  <a
    className="pokeCard__item "
    href="/"
    onClick={[Function]}
  >
    <h3
      className="pokeCard__name"
    >
       
      testing card
    </h3>
    <img
      alt="testing card"
      className="pokeCard__img"
      src="testingUrl"
    />
    <footer
      className="pokeCard__foot"
    />
  </a>
</styled.article>
`;
